File: scripts/changelog/split_archive.py
#!/usr/bin/env python3
"""
Split a large changelog archive into monthly files and normalize headings.
- Input: docs/changelog/Changelog-Archive-2025-08.md (default) or path arg
- Output: docs/changelog/Changelog-Archive-YYYY-MM.md per month
- Normalizes headings to: '## Version x.y.z'
- Ensures a '*Released: YYYY-MM-DD*' line if the date was only in the header
"""
import sys
import re
from pathlib import Path
from datetime import datetime

SRC = Path(sys.argv[1]) if len(sys.argv) > 1 else Path('docs/changelog/Changelog-Archive-2025-08.md')
OUT_DIR = Path('docs/changelog')

MONTHS = {
    'january': '01', 'february': '02', 'march': '03', 'april': '04', 'may': '05', 'june': '06',
    'july': '07', 'august': '08', 'september': '09', 'october': '10', 'november': '11', 'december': '12'
}

def month_name(ym: str) -> str:
    if ym == 'unknown':
        return 'Unknown'
    y, m = ym.split('-')
    dt = datetime(int(y), int(m), 1)
    return dt.strftime('%B %Y')

def parse_date(s: str):
    m = re.search(r'(\d{4})-(\d{2})-(\d{2})', s)
    if m:
        return f"{m.group(1)}-{m.group(2)}-{m.group(3)}"
    m = re.search(r'(January|February|March|April|May|June|July|August|September|October|November|December)\s+(\d{1,2}),\s*(\d{4})', s, re.I)
    if m:
        mon = MONTHS[m.group(1).lower()]
        day = str(int(m.group(2))).zfill(2)
        year = m.group(3)
        return f"{year}-{mon}-{day}"
    return None

def normalize_heading(header: str):
    # ## [0.2.6] - 2025-07-05
    m = re.match(r'^##\s*\[(\d+\.\d+\.\d+)\](?:\s*-\s*(\d{4}-\d{2}-\d{2}))?', header)
    if m:
        ver, date = m.group(1), m.group(2)
        return f"## Version {ver}", ver, date
    # ## Version 0.5.28 - "..."
    m = re.match(r'^##\s*Version\s*(\d+\.\d+\.\d+)', header)
    if m:
        ver = m.group(1)
        return f"## Version {ver}", ver, None
    return None, None, None

def ensure_released_line(lines, header_date):
    if not any(l.strip().startswith('*Released:') for l in lines) and header_date:
        return [lines[0], f"*Released: {header_date}*", *lines[1:]]
    return lines

def detect_month(lines, header_date):
    date = header_date
    if not date:
        for ln in lines:
            m = re.match(r'^\*Released:\s*(.+)\s*\*?\s*$', ln.strip())
            if m:
                d = parse_date(m.group(1))
                if d:
                    date = d
                    break
    if not date:
        for ln in lines:
            m = re.search(r'spans\s+(\d{4}-\d{2}-\d{2})[^\d]+(\d{4}-\d{2}-\d{2})', ln)
            if m:
                date = m.group(2)
                break
    if not date:
        for ln in lines:
            m = re.search(r'(\d{4}-\d{2}-\d{2})', ln)
            if m:
                date = m.group(1)
                break
    return (date[:7] if date else 'unknown')

def split_sections(text: str):
    lines = text.splitlines()
    # Find release section starts
    starts = [i for i, ln in enumerate(lines) if ln.startswith('## ')]
    sections = []
    for idx, start in enumerate(starts):
        end = starts[idx + 1] if idx + 1 < len(starts) else len(lines)
        header = lines[start]
        norm_header, ver, header_date = normalize_heading(header)
        if not norm_header:
            continue
        body = lines[start + 1:end]
        sec_lines = [norm_header, *body]
        sec_lines = ensure_released_line(sec_lines, header_date)
        sections.append((ver, header_date, sec_lines))
    return sections

def main():
    if not SRC.exists():
        print(f"Source archive not found: {SRC}", file=sys.stderr)
        sys.exit(1)
    text = SRC.read_text(encoding='utf-8')
    sections = split_sections(text)
    if not sections:
        print("No release sections found to split.", file=sys.stderr)
        sys.exit(1)
    buckets = {}
    for ver, header_date, lines in sections:
        ym = detect_month(lines, header_date)
        buckets.setdefault(ym, []).append('\n'.join(lines))
    # Write per-month files (overwrite)
    OUT_DIR.mkdir(parents=True, exist_ok=True)
    for ym, parts in buckets.items():
        out = OUT_DIR / f"Changelog-Archive-{ym}.md"
        label = month_name(ym)
        suffix = f" ({ym})" if ym != 'unknown' else ''
        header = f"# Changelog Archive — {label}{suffix}\n\nThis archive contains entries released in {label}.\n\n"
        out.write_text(header + '\n\n'.join(parts) + '\n', encoding='utf-8')
        print(f"Wrote {out}")

if __name__ == '__main__':
    main()

File: scripts/changelog/split-archive.js
#!/usr/bin/env node
/*
  Split a large changelog archive into monthly files and normalize headings.
  - Input: docs/changelog/Changelog-Archive-2025-08.md (default) or path arg
  - Output: docs/changelog/Changelog-Archive-YYYY-MM.md per month
  - Normalizes headings to: `## Version x.y.z`
  - Ensures a `*Released: YYYY-MM-DD*` line if the date was only in the header
*/
const fs = require('fs');
const path = require('path');

const SRC = process.argv[2] || path.join('docs', 'changelog', 'Changelog-Archive-2025-08.md');
const OUT_DIR = path.join('docs', 'changelog');

function monthName(ym) {
  const [y, m] = ym.split('-').map(Number);
  const date = new Date(Date.UTC(y, m - 1, 1));
  return date.toLocaleString('en-US', { month: 'long', year: 'numeric', timeZone: 'UTC' });
}

function parseDate(str) {
  // Try ISO first (YYYY-MM-DD)
  let m = str.match(/(\d{4})-(\d{2})-(\d{2})/);
  if (m) return `${m[1]}-${m[2]}-${m[3]}`;
  // Try Month DD, YYYY
  m = str.match(/(January|February|March|April|May|June|July|August|September|October|November|December)\s+(\d{1,2}),\s*(\d{4})/i);
  if (m) {
    const monthMap = {
      january: '01', february: '02', march: '03', april: '04', may: '05', june: '06',
      july: '07', august: '08', september: '09', october: '10', november: '11', december: '12'
    };
    const month = monthMap[m[1].toLowerCase()];
    const day = String(m[2]).padStart(2, '0');
    const year = m[3];
    return `${year}-${month}-${day}`;
  }
  return null;
}

function detectMonth(sectionLines, headerDate) {
  // Prefer header date if present
  let date = headerDate || null;
  // Search for *Released: ...
  if (!date) {
    for (const ln of sectionLines) {
      const m = ln.match(/^\*Released:\s*(.+)\s*\*?\s*$/);
      if (m) {
        const iso = parseDate(m[1]);
        if (iso) { date = iso; break; }
      }
    }
  }
  // Search for spans YYYY-MM-DD–YYYY-MM-DD -> pick end date
  if (!date) {
    for (const ln of sectionLines) {
      const m = ln.match(/spans\s+(\d{4}-\d{2}-\d{2})[^\d]+(\d{4}-\d{2}-\d{2})/i);
      if (m) { date = m[2]; break; }
    }
  }
  // Fallback: any ISO date in the section
  if (!date) {
    for (const ln of sectionLines) {
      const m = ln.match(/(\d{4}-\d{2}-\d{2})/);
      if (m) { date = m[1]; break; }
    }
  }
  if (!date) return 'unknown';
  return date.slice(0, 7); // YYYY-MM
}

function normalizeHeading(headerLine) {
  // Returns { header: '## Version x.y.z', version, headerDate }
  let version = null;
  let headerDate = null;
  let m;
  // ## [0.2.6] - 2025-07-05
  m = headerLine.match(/^##\s*\[(\d+\.\d+\.\d+)\](?:\s*-\s*(\d{4}-\d{2}-\d{2}))?/);
  if (m) {
    version = m[1];
    if (m[2]) headerDate = m[2];
    return { header: `## Version ${version}`, version, headerDate };
  }
  // ## Version 0.5.28 - "Tagline"
  m = headerLine.match(/^##\s*Version\s*(\d+\.\d+\.\d+)/);
  if (m) {
    version = m[1];
    return { header: `## Version ${version}`, version, headerDate };
  }
  return null;
}

function splitSections(text) {
  const lines = text.split(/\r?\n/);
  const sections = [];
  let i = 0;
  // Skip preface until first release heading
  while (i < lines.length && !/^##\s/.test(lines[i])) i++;
  while (i < lines.length) {
    // Find a release header
    while (i < lines.length && !/^##\s/.test(lines[i])) i++;
    if (i >= lines.length) break;
    const headerIdx = i;
    const headerLine = lines[i];
    const norm = normalizeHeading(headerLine);
    if (!norm) { i++; continue; }
    // Find next header start
    let j = i + 1;
    while (j < lines.length && !/^##\s/.test(lines[j])) j++;
    const section = lines.slice(headerIdx, j);
    sections.push({ headerLine, norm, lines: section });
    i = j;
  }
  return sections;
}

function ensureReleasedLine(sectionLines, headerDate) {
  const hasReleased = sectionLines.some(l => /^\*Released:/.test(l));
  if (!hasReleased && headerDate) {
    // Insert after header line
    const res = [...sectionLines];
    res.splice(1, 0, `*Released: ${headerDate}*`);
    return res;
  }
  return sectionLines;
}

function process() {
  if (!fs.existsSync(SRC)) {
    console.error(`Source archive not found: ${SRC}`);
    process.exit(1);
  }
  const raw = fs.readFileSync(SRC, 'utf8');
  const sections = splitSections(raw);
  if (!sections.length) {
    console.error('No release sections found to split.');
    process.exit(1);
  }
  const buckets = new Map(); // ym -> array of section strings
  for (const s of sections) {
    const { header, headerDate } = s.norm;
    // Normalize header
    const body = s.lines.slice(1);
    let outLines = [header, ...body];
    // Ensure Released line if we had a date in the header
    outLines = ensureReleasedLine(outLines, headerDate);
    const ym = detectMonth(outLines, headerDate);
    if (!buckets.has(ym)) buckets.set(ym, []);
    buckets.get(ym).push(outLines.join('\n'));
  }

  // Write each bucket to its own file
  for (const [ym, parts] of buckets.entries()) {
    const fname = `Changelog-Archive-${ym}.md`;
    const fpath = path.join(OUT_DIR, fname);
    const humanMonth = monthName(ym);
    const header = `# Changelog Archive — ${humanMonth} (${ym})\n\nThis archive contains entries released in ${humanMonth}.`;
    const content = header + '\n\n' + parts.join('\n\n');
    fs.writeFileSync(fpath, content, 'utf8');
    console.log(`Wrote ${fpath}`);
  }
}

process();

